### ISSUE TITLE
[Spike] Run containers-in-containers without chown'ing container's rootfs

### ISSUE THREAD

**Issue by taylorsilva** [concourse/concourse#6538]
## Summary

Currently baggageclaim does a [remapping](https://github.com/concourse/baggageclaim/blob/4b89831469e786e63d2290b12e60a031dfcf688f/uidgid/mapper_linux.go#L13) of a containers filesystem so it's owned by the container user. This ends up taking a while, therefore starting privileged containers is slow.

## Context

* Prior discussion: #6475

Prior testing we did:
```
- privileged true
- insde a user_namespace
- with root mapping to root
  - docker-compose does not work. fails mounting cgroups
  mount -n -t cgroup -o cpuset cgroup /sys/fs/cgroup/cpuset


- privileged true
- insde a user_namespace (owns all other namespaces)
- insde a cgroup_namespace
- root inside maps to max_id outside - baggageclaim does not need to chown/chmod the files
  - docker-compose fails with:
    mount: /sys/fs/cgroup: mount point not mounted or bad option.
    which may just be a scripting problem that could be fixed????
    maybe becuse we're in a cgroup namespace we don't need to call the "sanitize_cgroups" func in the docker-lib.sh script??
    next step: fly hijack the container and edit the script. manually go through starting docker again
  - modified script to not sanitize_cgroups() and docker started just fine
    - got error extracting layer when pulling postgres image. Could pull nginx image
    - running nginx image returned this error:
      applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/cpuset/docker: permission denied\"
    - file ownership is weird under /sys/. /sys/fs/cgroup is owned by root inside the container but all folders under cgroup are owned by "nobody" which is a user under /etc/passwd


- privileged true
- insde a user_namespace
- insde a cgroup_namespace
- with root mapping to root - baggageclaim needs to chown/chmod the files
    - docker-compose failed extracting a layer...
    ERROR: failed to register layer: Error processing tar file(exit status 1): operation not permitted
    - kept happening when extracting one layer in particular for the postgres image... and concourse/dev. Same happened while manually pulling the images after hijacking the container
```

**Comment by muntac** [comment]
Q: For non-privileged containers, does chown'ing happen??
A: chowns's some of the files (counted 1,875 out of 7828 files modified)
  sidenote: chmod's all chown'd files that are not symlinks (1,278)

Q: what happens when a run a privileged step? Do the files get chown/chmod'd?
A: chowns's some of the files (counted 2,359 out of 7828 files modified)
  sidenote: chmod's all chown'd files that are not symlinks (1,444)

The difference might initially seem small, about 400 more files chownd.
But for each chown OverlayFS will initiate a copy operation, which for large files
would be an expensive operation.

**Comment by muntac** [comment]
Exploring structure of `worker-state` to learn more about volumes.

Q: What are `overlays` and their directories on a worker??
A:
- `overlays` directory
  - at the root of this dir is the upperdir of cow volumes
  - there's a "work" dir which is something overlay needs... it's empty for reasons that overlay knows about and we don't
- `volumes` directory
  - there are three dir's that are for baggageclaim lifecycle purposes: init, live, dead
  - live dir:
    - contains all the overlay "merged" dir's

**Comment by muntac** [comment]
Q: What happens if the privilegedNamespacer in baggageclaim is a noop?
A:
- Most stuff works for the privileged container. Things break (like mounting) when the root user tries to run a binary with setuid enabled and its owned by a non-root user things don't work as expected.
- We can read/write files from outputs from non-privileged tasks just fine
  - these get remapped by the unprivilegedNamespacer still if they get sent back to a non-privileged container

**Comment by taylorsilva** [comment]
Measuring the initialization time when not mapping. We read the initialization time from the web UI.

tldr: the entire container initialization time is due to waking the entire rootfs + any chown'ing and chmod'ing

```
  ## Baseline - unmodified baggageclaim
      step type       | 1st Run   |   2nd Run   |   3rd Run   |   4th Run  | Average
  Non-privileged step | 10s       |  9s         |   11s       |   8s       |   9.5s
  Privileged step     | 2m6s      |  2m0s       |   2m6s      |   2m10s    |   2m5.5s
  Non-privileged step | 10s       |  12s        |   13s       |   8s       |   10.75s

  ## No-Mapping for privileged containers
      step type       | 1st Run   |   2nd Run   |   3rd Run   |   4th Run  | Average
  Non-privileged step | 12s       |   7s        |   8s        |   8s       |   8.75s   
  Privileged step     | 0s        |   1s        |   0s        |   1s       |   0.5s 
  Non-privileged step | 11s       |   7s        |   8s        |   8s       |   8.5s

## Privileged containers only have /bin/* chown'd and chmod'd
      step type       | 1st Run   |   2nd Run   |   3rd Run   |   4th Run  | Average
  Privileged step     | 1s        |   1s        |   1s        |   1s       |   1s 
```

**Comment by taylorsilva** [comment]
We changed the baggageclaim code so for privileged containers it would only chown files that have `setuid` set on their file. We ran a pipeline that builds an image and runs a Concourse inside a container using docker. These tasks both worked successfully and initialization times were around the same as non-privileged tasks.

```
## Privileged containers only have /bin/* chown'd and chmod'd
      step type       | 1st Run   |   2nd Run   |   3rd Run   |   4th Run  | Average
  Privileged step     | 7s        |   7s        |   7s        |   7s       |   7s 
```

**Comment by taylorsilva** [comment]
Running testflight with our modified baggageclaim, initialization is wayyyy down, to 8s
![image](https://user-images.githubusercontent.com/4583837/109532538-36977000-7a87-11eb-83c8-82cfaa925f1f.png)

In CI, initialization is betwen 2-3mins
![image](https://user-images.githubusercontent.com/4583837/109532655-53cc3e80-7a87-11eb-8b50-dec5fc2b9184.png)
